pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/ClusterChangeMediator.java

```diff
@@ -18,142 +18,139 @@
             break;
+          } catch (Exception e) {
+            LOGGER.error("Caught exception while handling changes", e);
           }
         }
-
-        LOGGER.warn("Stopping deferred cluster state update thread");
-        _deferredClusterUpdater = null;
       }
     };
+    _clusterChangeHandlingThread.start();
+  }
 
-    _deferredClusterUpdater.start();
+  private void processClusterChange(ChangeType changeType, ClusterChangeHandler changeHandler) {
+    long startTime = System.currentTimeMillis();
+    LOGGER.info("Start processing {} change", changeType);
+    changeHandler.processClusterChange();
+    long endTime = System.currentTimeMillis();
+    LOGGER.info("Finish processing {} change in {}ms", changeType, endTime - startTime);
+    _lastProcessTimeMap.put(changeType, endTime);
   }
 
   @Override
   public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
-    // If the deferred update thread is alive, defer the update
-    if (_deferredClusterUpdater != null && _deferredClusterUpdater.isAlive()) {
-      try {
-        _clusterChangeQueue.put(new ImmutablePair<>(UpdateType.EXTERNAL_VIEW, System.currentTimeMillis()));
-      } catch (InterruptedException e) {
-        LOGGER.warn("Was interrupted while trying to add external view change to queue", e);
-      }
-    } else {
-      LOGGER.warn(
-          "Deferred cluster updater thread is null or stopped, not deferring external view routing table rebuild");
-      _helixExternalViewBasedRouting.processExternalViewChange();
-      _tableQueryQuotaManager.processQueryQuotaChange();
-    }
+    Preconditions.checkState(externalViewList.isEmpty(), "Helix pre-fetch should be disabled");
+    Preconditions.checkState(_clusterChangeHandlingThread.isAlive(), "Cluster change handling thread is not alive");
+
+    enqueueChange(ChangeType.EXTERNAL_VIEW);
   }
 
   @Override
-  public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
-    // If the deferred update thread is alive, defer the update
-    if (_deferredClusterUpdater != null && _deferredClusterUpdater.isAlive()) {
-      try {
-        _clusterChangeQueue.put(new ImmutablePair<>(UpdateType.INSTANCE_CONFIG, System.currentTimeMillis()));
-      } catch (InterruptedException e) {
-        LOGGER.warn("Was interrupted while trying to add external view change to queue", e);
-      }
-    } else {
-      LOGGER.warn(
-          "Deferred cluster updater thread is null or stopped, not deferring instance config change notification");
-      _helixExternalViewBasedRouting.processInstanceConfigChange();
-    }
+  public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext changeContext) {
+    Preconditions.checkState(instanceConfigs.isEmpty(), "Helix pre-fetch should be disabled");
+    Preconditions.checkState(_clusterChangeHandlingThread.isAlive(), "Cluster change handling thread is not alive");
```

Comment:
Its not clear who handles this runtime exception and what the recovery path is. The earlier code processed the change inline.

Reply:
Changed the behavior to be: if thread is dead, log an error but process inline.

Reply:
Thanks. If this had gone uncaught, then even with the proactive mode enabled we wouldn't have been able to process the changes, right? Could you please note the testing that was done to validate the changes? If any reliable unit tests can be added to test the new code, that would be really good!

Reply:
That's true. And that's why the error logs and metrics are added. When we encounter issues next time, we can figure out the root cause by analyzing the logs and metrics.

Reply:
The intention was definitely not captured in the description. Regardless, given how critical processing EV changes is, we can always have failsafe mechanism with metrics and alerts helping us followup on when the issue happens.

Regardless, please include testing done (even if it is manual) so we can review the coverage the change has received.

Reply:
I am a little bit confused... This 2 lines have been changed to the old behavior if you check the updated code.